Incorporated wgodbout changes and added support for other file extensions

michaelqyee opened the pull request:
- wgodbout had suggestions to fix an overwriting bug, which is incorporated
- wgodbout also suggested to add organization by album, which is incorporated
- wgodbout suggested additions of error files vs. logging to standard out, which is incorporated
- added support for m4a, m4g, m4v, and wav extensions, which can be in iTunes libraries
- added improvement to filter file names by forbidden characters for FAT32, NTFS, and Mac OS

Commit:
Incorporated changes suggested by wgodbout and added support for file extensions beyond mp3

Tclack88 commented:
Thank you for the MR, I didn't see wgodbout's suggestions earlier. I will review the changes soon after I can perform some tests

Tclack88 commented:
There's a lot of good stuff in this new version, for instance it's waaaay faster without needing to convert all the files (4 minutes down to 10 seconds for test on my original data). But, one problem is a lot of the sorting doesn't happen, even on files that used to be mp3s (and as such don't need changing), are no longer being sorted into an artist folder. So as the original version leaves me with about 80 directories, this new version is leaving me with a mix of about 400 directories and files

Tclack88 reviewed (commented):
This works wonderfully; I cannot praise enough how useful it is to maintain the original file format, It dramatically drops runtime (and probably many people will like not losing audio quality when dropping to mp3). I will merge it however I have a few futher changes to make on top of it. If the song doesn't have an album, then the try block starting in 91 will not move the song. I have a lot of music that has no associated albums, so I'd like in those cases for them to be pushed into the artist directory. For this a nested try block will do:
(91-100 change to)
```python3
                        try:
                            album = re.sub(forbiddenregex, '', ''.join(song['album'])).strip()
                            artistalbum = os.path.join(artist, album)
                            if not os.path.isdir(artistalbum):
                                    os.mkdir(artistalbum)
                            shutil.move(i, artistalbum)
                        except: # if the song has no album, move to the artist dir
                            shutil.move(i, artist)
```

Additionally, you may notice above that line 97 is better as:
`artistalbum = os.path.join(artist, album)`, this maintains neutrality among OS's (Windows for example uses a "\").

If you see this message, feel free to make the changes, otherwise I will merge as is then make additional changes tomorrow.

Thanks again for the great input!

michaelqyee committed:
changes based on Tclack88 feedback

michaelqyee commented:
Thanks for the feedback. I added a new commit that incorporates your suggestion for using OS-agnostic path joining, and I also realized that you can create all of the directories in one go using `os.path.makedirs()`, so I went ahead and did that. I thought the behavior would be cleaner to nest everything at the same level, so instead of moving to the artist directory by default, I created a default name "Unknown Album" to nest the files under. Additionally, your review pointed out that if artist is not defined for a song, it will fail unexpectedly as well, so I added a default artist as well.

Tclack88 reviewed (approved):
Looks great and works great. Good call on the uknown artist as well. It's not something I dealt with in my original use case, but I'm sure it would affect others' aong libraries
